src/components/TaskclusterAuth/TaskclusterCallback.tsx

```diff
@@ -0,0 +1,129 @@
+  method: string;
+  body: string;
+  headers: {
+    'Content-Type': string;
+  };
+}
+
+// from treeherder
+const processErrorMessage = async (error: string, status) => {
+  let errorMessage = '';
+
+  if (status >= 500) {
+    errorMessage +=
+      'There was a problem retrieving the data. Please try again in a minute.';
+  }
+
+  if (status === 400) {
+    errorMessage += 'The action resulted in a bad request.';
+  }
+
+  if (error instanceof Object) {
+    const key = Object.keys(error);
+
+    errorMessage += ` ${key}: ${error[key]}`;
+  } else if (error) {
+    errorMessage += error;
+  }
+  return errorMessage ? errorMessage.trim() : error;
+};
+
+// from treeherder, TODO move this in another file/folder
+const getData = async (url: string, options: RequestOptions) => {
+  let failureStatus = null;
+  const response = await fetch(url, options);
+
+  if (!response.ok) {
+    failureStatus = response.status;
+  }
+
+  const contentType = response.headers.get('content-type');
+
+  if (contentType && contentType !== 'application/json' && failureStatus) {
+    const errorMessage = processErrorMessage(
+      `${failureStatus}: ${response.statusText}`,
+      failureStatus,
+    );
+    return { data: errorMessage, failureStatus };
+  }
+
+  let data = await response.json();
+
+  if (failureStatus) {
+    data = processErrorMessage(data, failureStatus);
+  }
+  return { data, failureStatus };
+};
+
+function TaskclusterCallback() {
+  const [searchParams] = useSearchParams();
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const code = searchParams.get('code') as string;
+  const state = searchParams.get('state') as string;
+  const requestState = localStorage.getItem('requestState');
+
+  const getCredentials = async (taskclusterCode: string) => {
+    const rootUrl = localStorage.getItem('tcRootUrl');
+
+    if (!rootUrl) {
+      setErrorMessage(errorMessage);
+      return;
+    }
+
+    // fetch token Bearer, it will be used to fetch the access_token
+    const options = {
+      mode: 'no-cors',
+      method: 'POST',
+      body: `grant_type=authorization_code&code=${taskclusterCode}&redirect_uri=${redirectURI}&client_id=${clientId}`,
+      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+    };
+    console.log(`${rootUrl}/login/oauth/token`, options);
+    const response = await getData(`${rootUrl}/login/oauth/token`, options);
+    console.log('response ', response);
+
+    if (response.failureStatus) {
+      setErrorMessage(`errorMessage ${response.data}`);
+      return;
+    }
+
+    // TODO continue with fetchCredentials using response access_token Bearer
+  };
+
+  useEffect(() => {
+    if (code && requestState && requestState === state) {
+      void getCredentials(code);
+    } else {
+      if (errorMessage) {
+        setErrorMessage(
+          errorMessage + `We received error: ${errorMessage} from Taskcluster.`,
+        );
+      }
+    }
+  }, [code, state, requestState]);
```

Comment:
My approach would be to throw a "real" Error when there is an error, and deal with it in an error boundary defined on the Route. See the documentation about that => https://reactrouter.com/en/main/route/error-element

It's pretty powerful because you can implement a completely different component to handle the error, possibly with a "try again" button or something like that. (for now just displaying the error would be good)

This makes it possible to focus on the happy case in the main component, but still handle errors separately.